src/main/java/org/shanerx/tradeshop/utils/Utils.java

```diff
@@ -476,55 +478,66 @@ public ExchangeStatus canExchangeAll(Shop shop, Inventory playerInv, int multipl
         return ExchangeStatus.SUCCESS; //Successfully completed trade
 	}
 
-    //Returns an arraylist of the itemstacks to be removed/added, if it could not get enough of an item, will return index 0 as null and index 1 as item it could not get enough of
-	public ArrayList<ItemStack> getItems(Inventory inventory, List<ShopItemStack> items, int multiplier) {
-        Inventory clone = Bukkit.createInventory(null, inventory.getStorageContents().length);
-        clone.setContents(inventory.getStorageContents());
-        ArrayList<ItemStack> ret = new ArrayList<ItemStack>();
-        int totalCount = 0, currentCount = 0;
-        debugger.log("ShopTradeListener > Inventory Type Being Searched: " + inventory.getType().name(), DebugLevels.TRADE);
-		debugger.log("ShopTradeListener > Inventory Location Being Searched: " + (inventory.getLocation() != null ? inventory.getLocation().toString() : "null"), DebugLevels.TRADE);
-
-		for (ShopItemStack item : items) {
-			totalCount += item.getAmount() * multiplier;
-			int count = item.getAmount() * multiplier;
+	//Returns an arraylist of the itemstacks to be removed/added, if it could not get enough of an item, will return index 0 as null and index 1 as item it could not get enough of
+	public ArrayList<ItemStack> getItems(Inventory inventory, List<ShopItemStack> search, int multiplier) {
+		Map<ItemStack, Integer> storage = new HashMap<>(), found = new HashMap<>();
+		ArrayList<ItemStack> good = new ArrayList<ItemStack>(), bad = new ArrayList<ItemStack>();
 
-			debugger.log("ShopTradeListener > Item Material Being Searched for: " + item.getItemStack().getType().name(), DebugLevels.TRADE);
-			debugger.log("ShopTradeListener > Item count: " + count, DebugLevels.TRADE);
+		debugger.log("ShopTradeListener > Search List: " + search, DebugLevels.TRADE);
 
-			for (ItemStack storageItem : clone.getStorageContents()) {
-				if (storageItem != null) {
-					boolean isSimilar = item.isSimilar(storageItem.clone());
-					if (isSimilar) {
+		bad.add(null);
```

Comment:
The code on the receiving end checks for `null` in the first position to determine if the return was bad or not and items in the positions(used to just check only the 1st) to tell the player what is missing. So this makes sure that if the mad list is returned the first item is always `null`

Reply:
Hmm this is kinda hacky, not sure I like it...

Reply:
In the code im working on now I moved it to be part of a createBadList method so there aren't weird add(null) calls, we could change to return a Tuple<Boolean, List> but it'll make things a lot more complicated since I'm making canExchangeAll pass the result along in a Tuple already for the better messaging. Changing would end in occasionally passing around a Tuple containing a Tuple which I like less then a predictable null value check(which has been there for a bit and hasn't thrown errors.

As a note the contents of the list are either the found items or the items that could not be found.